### PR TITLE
Update modeling for cattrs 25

### PR DESCRIPTION
**Pull request by mshriver**
Testing completed locally with archive, direct, and s3 uploads, using a Red Hat internal project that integrates pytest-ibutsu and injects run, result, and artifacts. 

## Summary by Sourcery

Adopt cattrs for structured serialization across the modeling, archiving, and sending layers; replace ad-hoc JSON serializers and print statements with cattrs unstructure hooks and logging; refine archiver and S3 uploader error handling; expand pytest plugin with report header and terminal summary hooks; tighten build configuration; and add extensive tests and documentation.

New Features:
- Replace custom JSON serializers with a cattrs converter and custom unstructure hooks for functions, methods, descriptors, builtins, exceptions, modules, and attrs classes
- Switch TestRun and TestResult to use cattrs unstructure in to_dict with automatic exclusion of private attributes
- Introduce structured logging in archiver, sender, s3_uploader, and pytest plugin in place of print statements
- Add pytest hooks for report header and terminal summary to display pytest-ibutsu status

Enhancements:
- Register global cattrs unstructure hook factories for attrs classes and non-serializable types and specific hooks for IbutsuTestRun and IbutsuTestResult
- Refactor archiver to use cattrs unstructure with fallback error payloads and unify artifact serialization
- Improve S3Uploader and IbutsuSender with enhanced error handling, logging, environment-driven configuration, and retry diagnostics
- Update pre-commit hooks, strict mypy settings, and CI workflow to enforce coverage and linting standards

Documentation:
- Add AGENTS.md with development tips, testing instructions, and CI usage guidelines

Tests:
- Add extensive tests covering modeling serialization (including nested and edge-case metadata), archiver, sender, s3_uploader, pytest plugin options and hooks, and full integration scenarios

### CODE REVIEW

File: tests/test_archive.py [hunk 1]
```diff
@@ -91,24 +95,31 @@ def test_data(
...
-        archives += 1 if re.match(ARCHIVE_REGEX, path.name) else 0
-    assert archives == 1, f"Expected exactly one archive file, got {archives}"
+    with subtests.test("file_creation_and_properties"):
+        # Test archive file was created with correct properties
+        archive_name = f"{run_id}.tar.gz"
+        archive = pytester.path.joinpath(archive_name)
+        assert archive.is_file()
+        assert archive.lstat().st_size > 0
+
+    with subtests.test("archive_count_validation"):
```
[the remark this answers is not in the document]
Comment (mshriver) [reply tests/test_archive.py:117]: yeah I'm not really a fan of the pattern anyway, might switch all these to parametrization.

File: tests/test_sender.py [hunk 2]
```diff
@@ -450,9 +468,12 @@ def test_retry_on_different_network_exceptions(
         mock_sleep.assert_called_once_with(1.0)
 
     @patch("pytest_ibutsu.sender.time.sleep")
```
[the remark this answers is not in the document]
Comment (mshriver) [reply tests/test_sender.py:470]: I'm going to leave a certain level of this in order to have the unit tests, even if their implementation is subpar. I'd rather have some level of coverage.

File: tests/test_sender.py [hunk 3]
```diff
@@ -247,8 +246,7 @@ def test_upload_artifact_under_limit(self, mock_print):
         args, kwargs = sender._make_call.call_args
         assert args[0] == sender.artifact_api.upload_artifact
 
-    @patch("builtins.print")
-    def test_upload_artifact_over_limit(self, mock_print):
+    def test_upload_artifact_over_limit(self, caplog):
```
[the remark this answers is not in the document]
Comment (mshriver) [reply tests/test_sender.py:249]: Done

**Comment by mshriver** [comment]
@sourcery-ai summary
